pkg/skuba/actions/auth/auth_test.go

```diff
@@ -0,0 +1,521 @@
+					<input tabindex="2" required id="password" name="password" type="password" class="theme-form-input" placeholder="password" />
+					</div>
+					<button tabindex="3" id="submit-login" type="submit" class="dex-btn theme-btn--primary">Login</button>
+				</form>
+				</div>
+				`, authReqID)
+			_, _ = w.Write([]byte(htmlOutput))
+		case http.MethodPost:
+			username := r.FormValue("login")
+			password := r.FormValue("password")
+			if username == MockDefaultUsername && password == MockDefaultPassword {
+				http.Redirect(w, r, fmt.Sprintf("%s/approval", url)+"?req="+authReqID, http.StatusSeeOther)
+			}
+		}
+	}
+}
+
+func TokenHandler() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		resp := struct {
+			AccessToken  string `json:"access_token"`
+			TokenType    string `json:"token_type"`
+			ExpiresIn    int    `json:"expires_in"`
+			RefreshToken string `json:"refresh_token,omitempty"`
+			IDToken      string `json:"id_token"`
+		}{
+			newID(),
+			"bearer",
+			86399,
+			MockRefreshToken,
+			MockIDToken,
+		}
+		data, _ := json.Marshal(resp)
+		w.Header().Set("Content-Type", "application/json")
+		w.Header().Set("Content-Length", strconv.Itoa(len(data)))
+		_, _ = w.Write(data)
+	}
+}
+
+func KeysHandler() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		key, _ := rsa.GenerateKey(rand.Reader, 1024)
+		jwk := jose.JSONWebKey{
+			Key:       key,
+			KeyID:     "keyId",
+			Algorithm: "RSA",
+		}
+
+		_ = json.NewEncoder(w).Encode(&map[string]interface{}{
+			"keys": []map[string]interface{}{{
+				"alg": jwk.Algorithm,
+				"kty": jwk.Algorithm,
+				"kid": jwk.KeyID,
+				"n":   n(&key.PublicKey),
+				"e":   e(&key.PublicKey),
+			}},
+		})
+	}
+}
+
+func ApprovalHandler() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		htmlOutput := fmt.Sprintf(`
+			<div class="theme-panel">
+			<h2 class="theme-heading">Login Successful</h2>
+			<p>Please copy this code, switch to your application and paste it there:</p>
+			<input type="text" class="theme-form-input" value="%s" />
+			</div>
+		`, newID())
+		_, _ = w.Write([]byte(htmlOutput))
+	}
+}
+
+func ApprovalInvalidBodyHandler() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		htmlOutput := fmt.Sprintf(`
+			<div class="theme-panel">
+			<h2 class="theme-heading">Login Successful</h2>
+			</div>
+		`)
+		_, _ = w.Write([]byte(htmlOutput))
+	}
+}
+
+func Test_doAuth(t *testing.T) {
+	tests := []struct {
+		name               string
+		srvCb              func() *httptest.Server
+		clientID           string
+		clientSecret       string
+		issuerURL          string
+		username           string
+		password           string
+		rootCAData         []byte
+		insecureSkipVerify bool
+		expectedError      bool
+	}{
+		{
+			name: "login with root CA provided",
```

Comment:
There is super set covered by Test_Login. When you test public interface, the internal function (doAuth in this case) will be covered naturally

Reply:
ya, I just want to separate black-box and white-box testing.

For black-box, it does not know the function `doAuth` will be executed.